pkg/kubectl/cmd/apply.go

```diff
@@ -613,12 +653,8 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 			}
 			return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf(createPatchErrFormat, original, modified, current), source, err)
 		}
-	case err != nil:
-		return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf("getting instance of versioned object for %v:", p.mapping.GroupVersionKind), source, err)
-	case err == nil:
-		// Compute a three way strategic merge patch to send to server.
-		patchType = types.StrategicMergePatchType
-		patch, err = strategicpatch.CreateThreeWayMergePatch(original, modified, current, versionedObject, p.overwrite)
+	case types.StrategicMergePatchType:
+		patch, err = strategicpatch.CreateThreeWayMergePatch(original, modified, current, lookupPatchMeta, p.overwrite)
```

Comment:
if this errors based on discovered schema info, should we fall back to the previous approach (compiled-in types or merge patch)?

Reply:
I think this PR already done what you suggest.
The flow is
1. Try to do discovery to get openapi schema by calling `openapiSchema, err = f.OpenAPISchema()`. If encounter error, don't return error, but set `openapiSchema` to `nil`.
1. Then try to lookup the resource by calling `schema = p.openapiSchema.LookupResource(p.mapping.GroupVersionKind)`.
1. If both of the above steps succeed, use openapi to compute patch. Otherwise, construct a versioned Object from scheme.
1. If encounter `NotRegisteredError`, fallback to JSON merge patch. If no error, use compiled-in types to build patch.

Reply:
@liggitt

Reply:
if 1 and 2 succeed, but the openapi schema contains directives the client does not understand and step 3 fails, this code hard fails and does not fall back to anything